Common/cpp/SharedItems/Shareables.cpp

```diff
@@ -221,11 +229,45 @@ jsi::Value ShareableObject::toJSValue(jsi::Runtime &rt) {
         rt, data_[i].first.c_str(), data_[i].second->getJSValue(rt));
   }
   if (nativeState_ != nullptr) {
-    obj.setNativeState(rt, nativeState_);
+    if (isNativeStateImplemented_ == IsNativeStateImplemented::Yes) {
+      obj.setNativeState(rt, nativeState_);
+    } else if (isNativeStateImplemented_ == IsNativeStateImplemented::Unknown) {
+      runWithNativeStateProbe(
+          [&]() { obj.setNativeState(rt, nativeState_); });
+    }
   }
   return obj;
 }
 
+void ShareableObject::makeNativeStateFromObject(
+    jsi::Runtime &rt,
+    const jsi::Object &object) {
+  if (object.hasNativeState(rt)) {
+    nativeState_ = object.getNativeState(rt);
+    isNativeStateImplemented_ = IsNativeStateImplemented::Yes;
+  }
+}
+
+void ShareableObject::makeNativeStateFromNativeStateSource(
+    jsi::Runtime &rt,
+    const jsi::Value &nativeStateSource) {
+  if (nativeStateSource.isObject() &&
+      nativeStateSource.asObject(rt).hasNativeState(rt)) {
+    nativeState_ = nativeStateSource.asObject(rt).getNativeState(rt);
+    isNativeStateImplemented_ = IsNativeStateImplemented::Yes;
+  }
+}
+
+void ShareableObject::runWithNativeStateProbe(
+    std::function<void()> &&block) {
+  try {
+    block();
+    isNativeStateImplemented_ = IsNativeStateImplemented::Yes;
+  } catch (...) {
```

Comment:
Does this 'Not implemented' exception have a specific type that we can implicitly catch here?